File: Fortran/2_ex/mtx_multi_vec_test.f90
program mtx_multi_vec_test

    !Undergraduate Student: Arturo Burgos
    !Professor: João Rodrigo Andrade
    !Federal University of Uberlândia - UFU, Fluid Mechanics Laboratory
    !MFLab, Block 5P, Uberlândia, MG, Brazil


    implicit none !To avoid variable definition problems, checking if they are correctly defined
    integer :: i
    real, DIMENSION(3,3) :: mtx1, mtx2, mtx3, mtx4, mtx5
    DOUBLE PRECISION :: start, finish
    
    call random_number(mtx1)
    call random_number(mtx2)
    

    !Here I initialize the first matrix
    print *, 'The random mtx1 matrix is : '

    write (*,1) '---------------- New Slice ------------------'
    1 format(a47)

    do i = 1,3
                    
        print*, mtx1(i,:)  
        
    end do

    write (*,2) '---------------- New Slice ------------------'
    2 format(a47)
    
    write(*,'(//A//)')

    !Here I initialize the second matrix
    print *, 'The random mtx2 matrix is : '

    write (*,3) '---------------- New Slice ------------------'
    3 format(a47)

    do i = 1,3
                    
        print*, mtx2(i,:)  
        
    end do

    write (*,4) '---------------- New Slice ------------------'
    4 format(a47)


    write(*,'(////A///)')


    !Here I print the multiplication matrix result point by point 
    !(the same as mtx1.*mtx2 in Julia or MATLAB)
    mtx3 = mtx1*mtx2
    
    print *, 'The point by point multiplication matrix result mtx3 is : '

    write (*,5) '---------------- New Slice ------------------'
    5 format(a47)

    do i = 1,3
                    
        print*, mtx3(i,:)  
        
    end do

    write (*,6) '---------------- New Slice ------------------'
    6 format(a47)

    write(*,'(////A///)')

    !Here I print the multiplication matrix result using matmul 
    !(the same as mtx1.mtx2 in Julia or MATLAB)
    mtx5 = matmul(mtx1,mtx2)

    print *, 'The point by point multiplication matrix result mtx5 is : '

    write (*,9) '---------------- New Slice ------------------'
    9 format(a47)

    do i = 1,3
                    
        print*, mtx5(i,:)  
        
    end do

    write (*,10) '---------------- New Slice ------------------'
    10 format(a47)

    call cpu_time(start)

    mtx4 = matmul(mtx1,mtx2)

    call cpu_time(finish)

    write(*,'(////A///)')

    print *, 'The multiplication matrix result mtx4 is : '

    write (*,7) '---------------- New Slice ------------------'
    7 format(a47)

    do i = 1,3
                    
        print*, mtx4(i,:)  
        
    end do

    write (*,8) '---------------- New Slice ------------------'
    8 format(a47)

    write(*,'(A/)') 
    print '("Elapsed time is: ",f6.5," seconds")', finish - start

end program mtx_multi_vec_test

File: Python/2_ex/mtx_multi_vec_test.py
# Undergraduate Student: Arturo Burgos
# Professor: João Rodrigo Andrade
# Federal University of Uberlândia - UFU, Fluid Mechanics Laboratory - MFLab, Block 5P, Uberlândia, MG, Brazil


# Second exercise: Consider two arbitrary Matrix NxN - by a common loop


import numpy as np  # Here I'm importing the array/matrix module from python
import time  # Here I'm importing this python module in order to calculate the precise time that the function works



n = 2  # The numbers of elements in each matrix dimension

A = np.random.randint(5, size = (n,n))  # Here I just import numpy to create an arbitrary matrix
B = np.random.randint(5, size = (n,n))

D = np.matmul(A,B) # Actual vectorizing function that does the same as the for loop below
R = np.zeros((n,n))

print('The matrix A is \n', A)
print('\n')
print('The matrix B is \n', B)
print('\n')
print('The matrix D is \n', D) # Print on the screen the result of the regular matmul function
# to compare both results
print('\n')

# i represents lines and j represents columns. k represents the line or column 11element11 indices

def multivec(mtx1, mtx2, z, mtx3):

    mtx3 = np.matmul(A,B)

    return mtx3

t_initial = time.time()  # I set here the initial time.
R = multivec(A, B, n, R)

print('\nThe result matrix of the multiplication is ')
print(R)

print('\n')
print("--- %s seconds ---\n" % (time.time() - t_initial))